### PR TITLE
Store certificate and path with well known identity

### PR DESCRIPTION
**Pull request by rnicoll**
This modifies Party to include the certificate path (and as part of that, the certificate for the
identity). This ensures certificate and path are stored for every well known identity, are available
for validation when registering identities.

### CODE REVIEW

**Comment by rnicoll** [comment]
Raising early for discussion - is this a sensible approach to having the certificate available? It feels sensible so far. Only real issue is... I presume nodes shouldn't actually have self-signed identities and that's why I can't build a valid path, right?

**Comment by mikehearn** [comment]
Haven't reviewed the code. The concept is fine although note that sometimes we may wish to have a Party without the backing proof of validity e.g. if the node sends a party to an RPC client, who cares, we can assume the node has checked it.

**Comment by rnicoll** [comment]
I'll try to break this down into small parts on Monday, but the API changes are at least done. Highlights:

* Use X509CertificateHolder in place of Certificate/X509Certificate/implementations thereof, so serialization is stable
* Add a root CA to the identity service and try to validate identities against that CA. At the moment this is null in many cases (at which point pretty much anything is accepted), but should be tightened up in M13.
* Node info and by extension registration now includes a certificate and path.
* Add storage and serialization handling for certificates and certificate paths

There's a bunch of TODOs and commented out tests in there, as well as a lack of documentation - I'll address on Monday, but right now would appreciate high level feedback.

File: node/src/main/kotlin/net/corda/node/services/identity/InMemoryIdentityService.kt [hunk 1]
```diff
@@ -26,27 +30,83 @@ import javax.annotation.concurrent.ThreadSafe
...
         principalToParties[party.name] = party
     }
 
+    @Throws(CertificateExpiredException::class, CertificateNotYetValidException::class, InvalidAlgorithmParameterException::class)
+    override fun registerAnonymousIdentity(anonymousParty: AnonymousParty, fullParty: PartyAndCertificate, path: CertPath) {
+        require(path.certificates.isNotEmpty()) { "Certificate path must contain at least one certificate" }
+        // Validate the chain first, before we do anything clever with it
+        val validator = CertPathValidator.getInstance("PKIX")
+        val validatorParameters = if (trustAnchor != null) {
+            PKIXParameters(setOf(trustAnchor))
```
Comment (mnesbit) [inline node/src/main/kotlin/net/corda/node/services/identity/InMemoryIdentityService.kt]: Bit unclear on this. We have a root certificate. The node config points at a truststore and also provides the password. Thus you can open the keystore. This contains all trusted certificate roots in the system.
Obviously this doesn't include the non-standard requirement that for any valid certificate chain at least one entry must be in the network map.

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt [hunk 2]
```diff
@@ -604,41 +606,30 @@ abstract class AbstractNode(open val configuration: NodeConfiguration,
         // TODO: Integrate with Key management service?
         val keyStore = KeyStoreWrapper(configuration.nodeKeystore, configuration.keyStorePassword)
         val privateKeyAlias = "$serviceId-private-key"
-        val privKeyFile = configuration.baseDirectory / privateKeyAlias
-        val pubIdentityFile = configuration.baseDirectory / "$serviceId-public"
-        val identityAndKey = keyStore.certificateAndKeyPair(privateKeyAlias)?.let { (cert, keyPair) ->
+        val privateKey = keyStore.certificateAndKeyPair(privateKeyAlias)
```
Comment (kchalkias) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: better rename this to sth like certAndKeyPair

File: node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt [hunk 3]
```diff
@@ -263,10 +263,9 @@ class ArtemisMessagingServer(override val config: NodeConfiguration,
         val trustStore = KeyStoreUtilities.loadKeyStore(config.trustStoreFile, config.trustStorePassword)
         val ourCertificate = keyStore.getX509Certificate(CORDA_CLIENT_TLS)
 
-        val ourSubjectDN = X500Name(ourCertificate.subjectDN.name)
         // This is a sanity check and should not fail unless things have been misconfigured
-        require(ourSubjectDN == config.myLegalName) {
-            "Legal name does not match with our subject CN: $ourSubjectDN"
+        require(ourCertificate.subject == config.myLegalName) {
+            "Legal name does not match with our subject CN: $ourCertificate.subject"
```
Comment (kchalkias) [inline node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt]: use ${}

File: core/src/main/kotlin/net/corda/core/crypto/Crypto.kt [hunk 4]
```diff
@@ -433,7 +438,7 @@ object Crypto {
         if (verificationResult) {
             return true
         } else {
-            throw SignatureException("Signature Verification failed!")
+            throw SignatureException("Signature verification failed for key ${publicKey.toStringShort()}")
```
Comment (rnicoll) [inline core/src/main/kotlin/net/corda/core/crypto/Crypto.kt]: Isn't the data going to be huge? Could print data hash, would that be acceptable

Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/crypto/Crypto.kt]: I was considering the Merkle root as clear data, this is what we practically sign, right?

File: core/src/test/kotlin/net/corda/core/crypto/X509UtilitiesTest.kt [hunk 5]
```diff
@@ -182,23 +189,23 @@ class X509UtilitiesTest {
...
+        sslCertAndKey.certificate.isValidOn(Date())
+        sslCertAndKey.certificate.isSignatureValid(JcaContentVerifierProviderBuilder().build(serverCertAndKey.certificate.subjectPublicKeyInfo))
 
-        assertTrue { sslCertAndKey.certificate.subjectDN.name.contains(MEGA_CORP.name.commonName) }
+        assertTrue { sslCertAndKey.certificate.subject.toString().contains(MEGA_CORP.name.commonName) }
         // Now sign something with private key and verify against certificate public key
         val testData = "123456".toByteArray()
         val signature = Crypto.doSign(DEFAULT_TLS_SIGNATURE_SCHEME, serverCertAndKey.keyPair.private, testData)
-        assertTrue { Crypto.isValid(DEFAULT_TLS_SIGNATURE_SCHEME, serverCertAndKey.certificate.publicKey, signature, testData) }
+        // assertTrue { Crypto.isValid(DEFAULT_TLS_SIGNATURE_SCHEME, serverCertAndKey.certificate.publicKey, signature, testData) }
```
Comment (mnesbit) [inline core/src/test/kotlin/net/corda/core/crypto/X509UtilitiesTest.kt]: I don't like leaving commented out tests. Either this is a TODO, or remove it

File: core/src/main/kotlin/net/corda/core/node/services/Services.kt [hunk 6]
```diff
@@ -396,7 +399,9 @@ interface KeyManagementService {
      * @return X.509 certificate and path to the trust root.
      */
     @Suspendable
-    fun freshKeyAndCert(identity: Party, revocationEnabled: Boolean): Pair<X509Certificate, CertPath>
+    fun freshKeyAndCert(identity: PartyAndCertificate, revocationEnabled: Boolean): Pair<X509CertificateHolder, CertPath>
+
+    fun getSigner(publicKey: PublicKey): ContentSigner
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/node/services/Services.kt]: I don't like leaking an object that can sign things. Would it be possible to have a getSignedCertificate method that works for our local keys, or something?

File: core/src/main/kotlin/net/corda/core/node/services/IdentityService.kt [hunk 7]
```diff
@@ -1,36 +1,43 @@
...
-    fun registerIdentity(party: Party)
+    /**
+     * Verify and then store a well known identity.
+     *
+     * @param party a party representing a legal entity.
+     * @throws IllegalArgumentException if the certificate path is invalid, or if there is already an existing
+     * certificate chain for the anonymous party.
+     */
+    @Throws(CertificateExpiredException::class, CertificateNotYetValidException::class, InvalidAlgorithmParameterException::class)
+    fun registerIdentity(party: PartyAndCertificate)
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/node/services/IdentityService.kt]: Should we add a verify method, which will check the path, etc? I know that right now you have not wired up the checks, but can we reduce the TODOs by pointing a lot of them at a dummy validation call?

File: core/src/test/kotlin/net/corda/core/crypto/X509UtilitiesTest.kt [hunk 8]
```diff
@@ -38,12 +41,12 @@ class X509UtilitiesTest {
...
-        caCert.checkValidity(Date()) // throws on verification problems
-        caCert.verify(caKey.public) // throws on verification problems
-        assertTrue { caCert.keyUsage[5] } // Bit 5 == keyCertSign according to ASN.1 spec (see full comment on KeyUsage property)
-        assertTrue { caCert.basicConstraints > 0 } // This returns the signing path length Would be -1 for non-CA certificate
+        assertTrue { caCert.subject.commonName == "Test Cert" } // using our subject common name
+        assertEquals(caCert.issuer, caCert.subject) //self-signed
+        caCert.isValidOn(Date()) // throws on verification problems
+        caCert.isSignatureValid(JcaContentVerifierProviderBuilder().build(caKey.public)) // throws on verification problems
+        // assertTrue { caCert.keyUsage[5] } // Bit 5 == keyCertSign according to ASN.1 spec (see full comment on KeyUsage property)
+        // assertTrue { caCert.basicConstraints > 0 } // This returns the signing path length Would be -1 for non-CA certificate
```
Comment (mnesbit) [inline core/src/test/kotlin/net/corda/core/crypto/X509UtilitiesTest.kt]: Is this a TODO, or a delete?

File: core/src/main/kotlin/net/corda/core/crypto/Crypto.kt (hunk 4 above)
Comment (rnicoll) [inline core/src/main/kotlin/net/corda/core/crypto/Crypto.kt]: It's what we sign for transactions, but there may be other use-cases. I'll have it log the data if it's short.

File: core/src/test/kotlin/net/corda/core/crypto/X509UtilitiesTest.kt (hunk 8 above)
Comment (rnicoll) [inline core/src/test/kotlin/net/corda/core/crypto/X509UtilitiesTest.kt]: It's a TODO, need to find if there's an equivalent for `X509CertificateHolder` or convert to `X509Certificate` and then run the test as normal.

File: core/src/main/kotlin/net/corda/core/crypto/Crypto.kt (hunk 4 above)
Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/crypto/Crypto.kt]: exactly, or truncate it to 40 chars max

**Comment by rnicoll** [comment]
I think that's everything addressed now

File: core/src/main/kotlin/net/corda/core/crypto/Crypto.kt [hunk 9]
```diff
@@ -436,7 +442,12 @@ object Crypto {
         if (verificationResult) {
             return true
         } else {
-            throw SignatureException("Signature Verification failed!")
+            val msg = if (clearData.size < THRESHOLD_INCLUDE_CLEAR_DATA) {
+                "Signature verification failed for key ${publicKey.toStringShort()}: ${clearData.toHex()}"
```
Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/crypto/Crypto.kt]: I guess it's fine for now, but we have to reconsider how big messages will be printed in the future. Eg truncate to first 40 chars then ...

**Pull request merged**